[509372] [HSD 15010715317] Add TBH dockerfile to manageability-docker folder and fix the configuration file malformed in inb container

yengliong93 opened the pull request:
_The PR review is to check for sustainability and correctness.  Sustainability is actually more business critical as correctness is largely tested into the code over time.   Its useful to keep in mind that SW often outlives the HW it was written for and engineers move from job to job so it is critical that code developed for Intel be supportable across many years.  It is up to the submitter and reviewer to look at the code from a perspective of what if we have to debug this 3 years from now after the author is no longer available and defect databases have been lost.  Yes, that happens all the time when we are working with time scales of more than 2 years.  When reviewing your code it is important to look at it from this perspective._

Author Mandatory (to be filled by PR Author/Submitter)
------------------------------------------------------
- Developer who submits the Pull Request for merge is required to mark the checklist below as applicable for the PR changes submitted.  
- Those checklist items which are not marked are considered as not applicable for the PR change.  
- Items marked with an asterisk suffix are mandatory items to check and if not marked will be treated as non-compliant pull requests by the developers for Inner Source Development Model (ISDM) compliance

### PULL DESCRIPTION
This PR uploads INBM dockerfile being used in TBH to manageability-docker folder.
ESH team can refer to this folder for the changes of TBH INBM dockerfile.

### REFERENCES
Reference URL for issue tracking (JIRA/HSD/Github): 
RTC:  https://rtc.intel.com/ccm0006001/web/projects/SSEA-%20IOTG%20PED%20Team#action=com.ibm.team.workitem.viewWorkItem&id=509372

HSD: https://hsdes.intel.com/appstore/article/#/15010715317

- [x] **_Added label to the Pull Request following the template: ISDM\_\<Complexity>\*_** \
	Note-1: Depending on complexity of code changes, use the suitable word for complexity: Low/Medium/High \
	Example: PR for Slim boot loader project with medium complexity can have the label as: ISDM_Medium	
- [x] Added label to the Pull Request for easier discoverability and search
- [x] RTC or HSD number will be included in final merge. HSD must always be included if available.
- N/A Changelogs are updated (or N/A if not customer visible)

### CODE MAINTAINABILITY
- [x] **_Commit Message meets guidelines as indicated in the URL\*_**
	- (https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md)
- [x] **_Every commit is a single defect fix and does not mix feature addition or changes\*_**
- N/A Added required new tests relevant to the changes
	- N/A PR contains URL links to functional tests executed with the new tests 
- N/A Updated Documentation as relevant to the changes
- N/A Updated Build steps/commands changes as relevant
- N/A PR change contains code related to security
- N/A PR introduces changes that breaks compatibility with other modules (If YES, please provide description)
- N/A Specific instructions or information for code reviewers (If any):
- N/A Run 'go fmt' or autopep8 as applicable.
- N/A New/modified methods and functions should have type annotations on signatures as applicable
- N/A New/modified methods must have appropriate doc strings (language dependent)

### APPLICATION SPECIFIC
- N/A Does PR change default config files under /etc? If so, will application still work after an upgrade that leaves /etc alone, like a Mender upgrade?
- N/A Is cloud UI changed? If so, are cloud definition files updated?



Maintainer Mandatory (to be filled by PR Reviewer/Approving Maintainer)
-----------------------------------------------------------------------
- Maintainer who approves the Pull Request for merge is required to mark the checklist below as appropriate for the PR change reviewed as key proof of attestation indicating reasons for merge. 
- Those checklist items which are not marked are considered as not applicable for the PR change. 
- Items marked with an asterisk suffix are mandatory items to check and if not marked will be treated as non-compliant pull requests by the maintainers for ISDM compliance.

### QUALITY CHECKS
- [ ] Architectural and Design Fit
- [x] **_Quality of code (At least one should be checked as applicable)\*_**
	- [x] Commit Message meets guidelines
	- [ ] PR changes adhere to industry practices and standards
	- [ ] Error and exception code paths implemented correctly
	- [ ] Code reviewed for domain or language specific anti-patterns
	- [ ] Code is adequately commented
	- [ ] Code copyright is correct
	- [ ] Confusing logic is explained in comments
	- [ ] Commit comment can be used to design a new test case for the changes
- [x] **_Test coverage shows adequate coverage with required CI functional tests pass on all supported platforms\*_**
- [x] **_Static code scan report shows zero critical issues\*_**
- [x] Integration tests are passing

### CODE REVIEW IMPACT
- Summary of Defects Detected in Code Review: **\<%P1*xx,P2*xx,P3*xx,P4*xx%>** \
Note P1/P2/P3/P4 denotes severity of defects found (Showstopper/High/Medium/Low) and xx denotes number of defects found

### SECURITY CHECKS
Please check if your PR fulfills the following requirements:

- [ ] Follow best practices when handling primitive data types
- [ ] Configure minimal permissions when opening pipes and ports
- [ ] Check contents within input structures are valid before use
- [ ] All forms of input validated
- [ ] Avoid inter-process race conditions
- [ ] Error and exception handling implemented
- [ ] Defend against Canonical Representation Issues - Any paths utilized?
- [ ] Follow 'secure by default' - Any configuration values added
- [ ] Fail safe - Any failure scenarios?
- [ ] Clean up temporary files - Any temporary files being used?

# _Code must act as a teacher for future developers_

yengliong93 commented:
@nmgaston @gblewis1  @tsirlapu  This PR added TBH dockerfile to manageability-docker/TBH folder. Please have a review, thanks.

nmgaston commented on manageability-docker/TBH/LICENSE:
```diff
@@ -0,0 +1,198 @@
+Press 'q' to quit once you have read the license.
```
This looks like the proprietary license.  We need to use the Apache 2.0 license.  Did you find this one somewhere else in our code?  If so, we need to exchange it with the opensource one.  You can find the Apache 2.0 one at the root level.

nmgaston commented on manageability-docker/TBH/Release-README:
```diff
@@ -0,0 +1,24 @@
+QUICK README
+=============
+1.  To install both TC and BC run the install-inb.sh script with root privileges. Before running the script update the following fields in azure_conf_file with the necessary information to have device connected to cloud.
```
We need to remove references to TC and BC.  Just use INBM or Intel In-band Manageability instead.

nmgaston commented on manageability-docker/TBH/Release-README:
```diff
@@ -0,0 +1,24 @@
+QUICK README
+=============
+1.  To install both TC and BC run the install-inb.sh script with root privileges. Before running the script update the following fields in azure_conf_file with the necessary information to have device connected to cloud.
+        - SCOPE_ID <refer Azure user guide documentation on how to fetch SCOPE_ID>
```
Actually we need to just point them to the Install guides in our /doc directory.  Did you find this somewhere else in our code?  We need to update that if so.

nmgaston commented on manageability-docker/TBH/inbm-third-party-programs.txt:
```diff
@@ -0,0 +1,1794 @@
+Turtle Creek Third Party Program Files
```
We should be using the third-party-programs.txt file only.  We are only using one file now for the product as a whole instead of 2 separate since the Open source.  https://github.com/intel/intel-inb-manageability/blob/develop/third-party-programs.txt

nmgaston commented on manageability-docker/TBH/inbm-vision-third-party-programs.txt:
```diff
@@ -0,0 +1,610 @@
+Bit Creek Third Party Program Files
```
Can remove this one and just use the third-party-programs.txt file only.  We are only using one file now for the product as a whole instead of 2 separate since the Open source.  https://github.com/intel/intel-inb-manageability/blob/develop/third-party-programs.txt

nmgaston commented on manageability-docker/TBH/Dockerfile:
```diff
@@ -0,0 +1,91 @@
+RUN dpkg -i ./trtl*.deb
+#RUN sed -i 's/Wants/#Wants/g' /etc/systemd/system/mqtt.service
+#RUN sed -i 's/After/#After/g' /etc/systemd/system/mqtt.service
+RUN mkdir -p /lib/firmware/intel-flashless
```
Will we need these commented out lines in the future or can they be removed?

gblewis1 replied:
I would say remove commented out lines

gblewis1 commented:
@yengliong93 Could this be done with a single Dockerfile? Docker is designed to allow containers to be platform-independent, and be customized at runtime as part of docker run. Maintaining multiple Dockerfiles increases the chance for errors.

gblewis1 commented on manageability-docker/TBH/Dockerfile:
```diff
@@ -0,0 +1,91 @@
+FROM ubuntu:20.04 as base
```
I had some problems using ubuntu:20.04 as base--suggest jrei/systemd-ubuntu:20.04

gblewis1 commented on manageability-docker/TBH/Dockerfile:
```diff
@@ -0,0 +1,91 @@
+RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y init mosquitto cryptsetup openssl dmidecode dbus vim net-tools rsyslog libffi-dev pciutils lxc-common dkms libssl-dev libpython3.8 ca-certificates wget openssh-server && \
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/*
+RUN find /etc/systemd/system \
```
this RUN command can be entirely removed with jrei/systemd-ubuntu:20.04

gblewis1 commented on manageability-docker/TBH/Dockerfile:
```diff
@@ -0,0 +1,91 @@
+RUN systemctl set-default multi-user.target
+
+
+STOPSIGNAL SIGRTMIN+3
```
What is this signal for?

gblewis1 commented on manageability-docker/TBH/Dockerfile:
```diff
@@ -0,0 +1,91 @@
+RUN mkdir -p /cache
+RUN dpkg -i ./inbc-program*.deb
+RUN dpkg -i ./inbm-vision-agent-*.deb
+#Need to fix: the mqtt.service file not able to load the mqtt-broker.key file from /etc/intel-manageability/secret/mqtt-broker dir.
```
Has this been fixed?

gblewis1 commented on manageability-docker/TBH/Dockerfile:
```diff
@@ -0,0 +1,91 @@
+RUN groupadd -f vision-agent
+RUN groupadd -f inbc-program
+RUN mkdir -p /var/persistent-log/mosquitto
+VOLUME ["/var/persistent-log/mosquitto"]
```
Why VOLUME here and not in the main Dockerfile?

gblewis1 commented on manageability-docker/TBH/Dockerfile:
```diff
@@ -0,0 +1,91 @@
+RUN chgrp mqtt-broker /var/persistent-log/mosquitto
+RUN chmod g+rwx /var/persistent-log/mosquitto
+RUN chmod -R g+rw /var/persistent-log/mosquitto
+VOLUME ["/lib/firmware/flashless_backup"]
```
Why VOLUME here and not in the main Dockerfile?

gblewis1 commented on manageability-docker/TBH/cloud_source:
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+set -eo pipefail
+
+cp /src/adapter.cfg /etc/intel-manageability/secret/cloudadapter-agent/adapter.cfg 
```
Will this leave behind a secret in /src?

gblewis1 commented on manageability-docker/TBH/docker-ble-policy:
```diff
@@ -0,0 +1,47 @@
+#include <tunables/global>
+
+profile docker-ble-policy flags=(attach_disconnected,mediate_deleted) {
```
change docker-ble-policy -> docker-manageability-policy (ble stands for bluetooth low energy)

gblewis1 commented on manageability-docker/TBH/run.sh:
```diff
@@ -0,0 +1,138 @@
+  echo "Run inb container in detach mode."
+  docker-compose up -d
+  sleep 30
+  echo "Stop inb container."
```
This looks like debug code.

gblewis1 commented on manageability-docker/TBH/run.sh:
```diff
@@ -0,0 +1,138 @@
+                ;;
+
+        *)
+                #echo "What would you like to do? $0 {--build | --deploy [autonomous/bypass/streaming] | --stop}"
```
the deploy option doesn't look at parameters autonomous/bypass/streaming

gblewis1 commented on manageability-docker/TBH/setup:
```diff
@@ -0,0 +1,17 @@
+#!/bin/sh
```
This file shouldn't be needed if using jrei/systemd-ubuntu:20.04  -- see https://hub.docker.com/r/jrei/systemd-ubuntu

yengliong93 commented:
> 
Hi @gblewis1 
Initially, I have thought the same thing. After that, I found there are some differences in Dockerfile and docker-compose.yml. 
For example, in TBH dockerfile, it installs pms-host package and links RM daemon. 

In docker-compose,yml, some directories and libraries are mounted specifically for TBH. 
For example:
/lib/firmware
/opt/xlink_provision:/opt/xlink_provision
/usr/lib/libXLink.so
/usr/lib/libSecureXLink.so
/usr/lib64/libSecureXLink_provision.so

nmgaston commented on manageability-docker/vision-docker/docker-compose.yml:
```diff
@@ -0,0 +1,47 @@
+version: "3.8"
+# Copyright: Copyright (C) 2021 Intel Corporation
```
could you add 2021-2022 to the copyright.

yengliong93 commented:
Updated LICENSE file to use Apache 2.0 license

Updated Release-README to point to Installation Guide on INBM open source.
https://github.com/intel/intel-inb-manageability/blob/develop/docs/In-Band%20Manageability%20Installation%20Guide%20Ubuntu.md

Remove inbm-vision-third-party-programs and inbm-third-party-programs text files, added third-party-programs text file.

Remove setup

Rename docker-ble-policy to docker-manageability-policy

Update run.sh to address the comment.

nmgaston commented on manageability-docker/vision-docker/third-party-programs.txt:
```diff
@@ -0,0 +1,1083 @@
+Intel In-Band Manageability Third Party Program Files
```
I should have thought about this more on my original comment.  We shouldn't have duplicate files since one will get updated and not the other.  Could we link to the one at root instead?

yengliong93 commented:
- Update docker-compose file.
- Link third-party-programs.txt from root to manageability-docker/vision-docker/third-party-programs.txt

yengliong93 commented:
- Updated Dockerfile to remove the volume declaration. Instead of declaring the volume in Dockerfile, the volume has been declared in docker-compose file.

nmgaston commented on manageability-docker/vision-docker/third-party-programs.txt:
```diff
@@ -0,0 +1 @@
+../../third-party-programs.txt
```
@gblewis1 - Do you have any recommendations on how we can include the third-party-programs.txt that we have at root with the docker files?  I don't know that this will do it.  We don't want to have duplicate documents.  Maybe just having it at the root level is enough.  What do you think?

nmgaston replied:
I talked to Gavin and we are just going to remove this since these aren't packaged.  Having the file at the root level is all we need.  Otherwise, it will cause mismatches.

nmgaston commented on manageability-docker/vision-docker/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+FROM jrei/systemd-ubuntu:20.04 as base
```
@gblewis1 - Do you think we should have the copyright added to all of these Docker related files?  And scripts?

gblewis1 replied:
@nmgaston We could do that.. maybe SPDX identifier would be best.

gblewis1 replied:
I talked to Natalie and Yeng Liong and we'll do this in another PR.

gblewis1 commented on manageability-docker/vision-docker/docker-manageability-policy:
```diff
@@ -0,0 +1,47 @@
+#include <tunables/global>
+
+profile docker-ble-policy flags=(attach_disconnected,mediate_deleted) {
```
docker-ble-policy -> docker-manageability-policy on this line

gblewis1 commented on manageability-docker/vision-docker/docker-manageability-policy:
```diff
@@ -0,0 +1,47 @@
+  mount,
+  umount,
+
+  signal (send,receive) peer=docker-ble-policy,
```
-ble- -> -manageability-

gblewis1 commented on manageability-docker/vision-docker/docker-manageability-policy:
```diff
@@ -0,0 +1,47 @@
+
+
+  # suppress ptrace denials when using 'docker ps' or using 'ps' inside a container
+  ptrace (trace,read,tracedby,readby) peer=docker-ble-policy,
```
-ble- -> -manageability-

gblewis1 commented on manageability-docker/vision-docker/run.sh:
```diff
@@ -0,0 +1,135 @@
+#!/bin/bash
+set -eox pipefail
+#sudo apparmor_parser -r -W docker-ble-policy
```
remove commented line or at least rename -ble- -> -manageability-

yengliong93 commented:
- LICENSE file removed. 
- All -ble- changed to -manageability- .

gblewis1 commented:
@nmgaston tests are passing. can you review/approve?
